### PR TITLE
Verify signature_or_mac_3

### PR DESCRIPTION
**Pull request by chrysn**
There are some minor fixes, but primarily this introduces a check for the signature / MAC on message 3. (This is based on #17 and #18, so please ignore these first commits; this may be easier to process if they are merged first).

@TimothyClaeys please check whether the basic operation fits in your concepts for py-edhoc.

This can be tested by checking out the edhoc branch of aiocoap and running the `edhocserver.py` script and `edhocclient.py localhost`, possibly with a `--static-static` option.

### CODE REVIEW

**Comment by chrysn** [comment]
Procedurally, it may also make sense to merge the fixes first (leaving out the last commit) and for me to merge that later with the checks on signature_or_mac2.

**Comment by chrysn** [comment]
Now extended to also verify mac_or_signature2; it'll probably need a round of refactoring to pass the tests again.

**Comment by chrysn** [comment]
Changes (including a consistent view on id_cred_remote) are regrouped as planned now, with a single "verify" commit on top.

Tests start breaking at "cred / authkey: Evaluate callback eagerly, deduplicating code"; I'd need a bit of help understanding what fails here: if you check what's happening at tests/test_initiator.py line 26, in some cases cred_idr has not been set yet at all, so the initiator.remote_authkey would not have contained anything sensible before (None rather than what it now does in raising an AttributeError, which IMO is the better behavior for "not having been set"). How can that have yielded any sensible data before?

(Along with that, I think this is now ready for review).

**Comment by chrysn** [comment]
Tests are now all fixed, most of them in a manner I'd consider compatible (in the sense that the fixes are really to the test system, either because parts of it were buggy and just didn't trigger, or because they represent a usage pattern that is unobtainable outside test settings).

What was not fixed that cleanly was testing of the mac_or_signature, as the auth keys for cred_type 0 (what *is* that? don't find it in the documented parts) are simply absent (None). In these cases, a warning is issued and the tests are aborted prematurely.

With this, the test suite passes, I think this is good to review and merge.

**Comment by TimothyClaeys** [comment]
Thanks a lot for this update. I've went over it quickly (will do another more thorough pass later today) and it looks really good. The idea behind the `cred_type` parameter was to distinguish between the different credential types (either RPK, C.509 or X.509) when providing them to the Initiator/Responder, and based on that value change the behaviour of the tests. As you mentioned, it is not something from the standard but I added it to make my life simpler. I'm not sure if it is still necessary.

**Review by TimothyClaeys — APPROVED** [review]

File: edhoc/roles/responder.py [hunk 1]
```diff
@@ -236,6 +236,23 @@ def finalize(self, message_three: bytes) -> Union[Tuple[bytes, bytes, int, int],
...
+        mac_3 = self._mac(self.cred_idi, self.remote_cred, self._hkdf3, 'K_3m', 16, 'IV_3m', 13, self._th3_input, self._prk4x3m, self.aad3_cb)
+
+        if not self.is_static_dh(self.remote_role):
+            external_aad = self._external_aad(self.remote_cred, self._th3_input, self.aad3_cb)
+            cose_sign = Sign1Message(
+                phdr=self.cred_idi,
+                uhdr={headers.Algorithm: self.cipher_suite.sign_alg},
+                payload=mac_3,
+                external_aad=external_aad)
+            # FIXME peeking into internals (probably best resolved at pycose level)
```
Comment (TimothyClaeys) [inline edhoc/roles/responder.py:249]: Yes I agree, using the `pycose `library here in `edhoc `has uncovered some design flaws in the API and its usage.

**Comment by TimothyClaeys** [comment]
Merging in the changes. CI will fail because of the newer pycose package.

**Pull request merged**